polkadot/collator/src/lib.rs

```diff
@@ -117,30 +133,148 @@ pub fn collate_ingress<'a, R>(relay_context: R)
 	where
-		R: RelayChainContext,
-	    R::Error: 'a,
+		R: RelayChainContext + 'a,
+		R::Error: 'a,
 		R::FutureEgress: 'a,
 		P: ParachainContext + 'a,
 {
-	Box::new(collate_ingress(relay_context).map(move |ingress| {
-		let (block_data, _, signature) = para_context.produce_candidate(
+	collate_ingress(relay_context).map(move |ingress| {
+		let (block_data, head_data) = para_context.produce_candidate(
+			last_head,
 			ingress.0.iter().flat_map(|&(id, ref msgs)| msgs.iter().cloned().map(move |msg| (id, msg)))
 		);
 
-		parachain::Candidate {
+		let signature = key.sign(&block_data.0[..]).into();
+		let pubkey_bytes: [u8; 32] = key.public().into();
+
+		let receipt = parachain::CandidateReceipt {
 			parachain_index: local_id,
-			collator_signature: signature,
-			block: block_data,
-			unprocessed_ingress: ingress,
+			collator: pubkey_bytes.into(),
+			signature,
+			head_data,
+			balance_uploads: Vec::new(),
+			egress_queue_roots: Vec::new(),
+			fees: 0,
+		};
+
+		parachain::Collation {
+			receipt,
+			block_data,
+		}
+	})
+}
+
+/// Polkadot-api context.
+struct ApiContext;
+
+impl RelayChainContext for ApiContext {
+	type Error = ();
+	type FutureEgress = Result<Vec<Vec<Message>>, Self::Error>;
+
+	fn routing_parachains(&self) -> BTreeSet<ParaId> {
+		BTreeSet::new()
+	}
+
+	fn unrouted_egress(&self, _id: ParaId) -> Self::FutureEgress {
+		Ok(Vec::new())
+	}
+}
+
+struct CollationNode<P, E> {
+	parachain_context: P,
+	exit: Option<E>,
+	para_id: ParaId,
+	key: Arc<ed25519::Pair>,
+}
+
+impl<P, E> Worker for CollationNode<P, E> where
+	P: ParachainContext + 'static,
+	E: Future<Item=(),Error=()> + Send + 'static
+{
+	type Work = Box<Future<Item=(),Error=()>>;
+	type Exit = future::Either<E, future::Empty<(),()>>;
+
+	fn exit(&mut self) -> Self::Exit {
+		match self.exit.take() {
+			Some(exit) => future::Either::A(exit),
+			None => future::Either::B(future::empty()),
 		}
-	}))
+	}
+
+	fn work<C: ServiceComponents>(self, service: &Service<C>) -> Self::Work
+		where ClientError: From<<<<C as ServiceComponents>::Backend as ClientBackend<PolkadotBlock>>::State as StateMachineBackend>::Error>,
+	{
+		let CollationNode { parachain_context, exit, para_id, key } = self;
+		let _ = exit;
+		let client = service.client();
+		let api = service.api();
+
+		let work = client.import_notification_stream()
+			.and_then(move |notification| {
+				let id = BlockId::hash(notification.hash);
+
+				match api.parachain_head(&id, para_id) {
+					Ok(Some(last_head)) => {
+						let collation_work = collate(
+							para_id,
+							HeadData(last_head),
+							ApiContext,
+							parachain_context.clone(),
+							key.clone(),
+						).map(Some);
+
+						future::Either::A(collation_work)
+					}
+					Ok(None) => {
+						info!("Parachain {:?} appears to be inactive. Cannot collate.", id);
+						future::Either::B(future::ok(None))
+					}
+					Err(e) => {
+						info!("Could not collate for parachain {:?}: {:?}", id, e);
```

Comment:
shouldn't that be at least of level `warn!`? I feel consuming an error without actually recovering and not being verbose about it, could easily cover up important issues to be addressed.

Reply:
it might get rather spammy, but yes